dbt/adapters/databricks/impl.py

```diff
@@ -214,6 +225,74 @@ def _get_columns_for_catalog(self, relation: DatabricksRelation) -> Iterable[Dic
     def valid_incremental_strategies(self) -> List[str]:
         return ["append", "merge", "insert_overwrite"]
 
+    @available.parse_none
+    @log_code_execution
+    def submit_python_job(
+        self, parsed_model: dict, compiled_code: str, timeout: Optional[int] = None
+    ) -> AdapterResponse:
+        # TODO limit this function to run only when doing the materialization of python nodes
+
+        credentials: DatabricksCredentials = self.config.credentials
+
+        if not timeout:
+            timeout = 60 * 60 * 24
+        if timeout <= 0:
+            raise ValueError("Timeout must larger than 0")
+
+        command_name = f"dbt-databricks_{version}"
+
+        invocation_env = os.environ.get(DBT_DATABRICKS_INVOCATION_ENV)
+        if invocation_env is not None and len(invocation_env) > 0:
+            self.ConnectionManager.validate_invocation_env(invocation_env)
+            command_name = f"{command_name}-{invocation_env}"
+
+        command_name += "-" + str(uuid.uuid1())
+
+        api_client = Api12Client(
+            host=credentials.host, token=cast(str, credentials.token), command_name=command_name
+        )
+
+        cluster_id = credentials.cluster_id
+        if not cluster_id:
+            raise ValueError("Python model doesn't support SQL Warehouses")
+
+        try:
+            # Create an execution context
+            context_id = api_client.Context.create(cluster_id)
+
+            try:
+                # Run a command
+                command_id = api_client.Command.execute(
+                    cluster_id=cluster_id,
+                    context_id=context_id,
+                    command=compiled_code,
+                )
+
+                # poll until job finish
+                status = None
+                start = time.time()
+                terminal_states = ("Cancelled", "Error", "Finished")
+                while status not in terminal_states and time.time() - start < timeout:
+                    time.sleep(1)
+                    json_resp = api_client.Command.status(
+                        cluster_id=cluster_id, context_id=context_id, command_id=command_id
+                    )
+                    status = json_resp["status"]
+                    # logger.debug(f"Polling.... in state: {state}")
+                if status != "Finished":
+                    raise dbt.exceptions.RuntimeException(
+                        "python model run ended in state"
+                        f"{status} with state_message\n{json_resp['results']['data']}"
```

Comment:
I noticed that if the python code actually failed, status will still be finish, but `json_resp['results']['resultType']` will be `error`, and `json_resp['results']['cause']` will contain traceback. I reused your code for jobs API and pushed the changes [here](https://github.com/dbt-labs/dbt-spark/pull/442/files). Feel free to review and let me know if any adjustment needed!

Reply:
Thanks! I'll address it.